ssl/ssl_ciph.c

```diff
@@ -1469,6 +1471,15 @@ STACK_OF(SSL_CIPHER) *ssl_create_cipher_list(const SSL_METHOD *ssl_method, STACK
         return NULL;
     }
 
+    /* Add TLSv1.3 ciphers first - we always prefer those if possible */
+    for (i = 0; i < sk_SSL_CIPHER_num(tls13_ciphersuites); i++) {
+        if (!sk_SSL_CIPHER_push(cipherstack,
+                                sk_SSL_CIPHER_value(tls13_ciphersuites, i))) {
+            sk_SSL_CIPHER_free(cipherstack);
+            return NULL;
+        }
+    }
+
```

Comment:
I see the TLS 1.3 ciphers go on top.  How unusual do you think it will be for users to build OpenSSL 1.1.1 with RC4 and TLS 1.0 still enabled?  Some rather dated TLS 1.0 libraries only see the first 64 ciphers in the client handshake, and only support TLS 1.0 with RC4 and 3DES.  When these get pushed further down the list, they become invisible to these (Microsoft Windows 2003 Server most likely) implementations. So one might argue that the TLS 1.3 ciphers should go last (since the protocol is chosen first, this won't matter to real TLS 1.3 peers).  On the other hand, do we just say good-riddance to interoperability with 15-year old TLS 1.0 stacks with only ciphers we no longer enable by default?

Reply:
Of course users who desperately want interop, can tweak the TLS 1.2 cipher list to make sure that RC4 (at least for Exchange2003 3DES is broken, and only RC4 works) is above slot 62 (rather than 65 as before)

Reply:
> I see the TLS 1.3 ciphers go on top. How unusual do you think it will be for users to build OpenSSL 1.1.1 with RC4 and TLS 1.0 still enabled? Some rather dated TLS 1.0 libraries only see the first 64 ciphers in the client handshake, and only support TLS 1.0 with RC4 and 3DES.

By default we're only sending through 3 TLSv1.3 ciphersuites, so that still leaves the next 61 ciphersuites for a match to be found for these old (buggy) libraries. Personally I don't think we should worry about this.